odbc/src/tests.rs

```diff
@@ -0,0 +1,244 @@
+                std::mem::transmute::<*mut *mut MongoHandle, *mut Handle>(handle_ptr),
+            )
+        );
+        assert_eq!(
+            ConnectionState::Allocated,
+            (*handle).as_connection().unwrap().read().unwrap().state
+        );
+        assert_eq!(
+            1,
+            (*env_handle)
+                .as_env()
+                .unwrap()
+                .read()
+                .unwrap()
+                .connections
+                .len()
+        );
+        assert_eq!(
+            EnvState::ConnectionAllocated,
+            (*env_handle).as_env().unwrap().read().unwrap().state
+        );
+        assert_eq!(
+            SqlReturn::SUCCESS,
+            SQLFreeHandle(
+                HandleType::Dbc,
+                std::mem::transmute::<*mut MongoHandle, Handle>(handle),
+            )
+        );
+        assert_eq!(
+            0,
+            (*env_handle)
+                .as_env()
+                .unwrap()
+                .read()
+                .unwrap()
+                .connections
+                .len()
+        );
+        assert_eq!(
+            EnvState::Allocated,
+            (*env_handle).as_env().unwrap().read().unwrap().state
+        );
+    }
+}
+
+#[test]
+fn statement_alloc_free() {
+    unsafe {
+        let env_handle: *mut _ =
+            &mut MongoHandle::Env(RwLock::new(Env::with_state(EnvState::Allocated)));
+
+        let conn_handle: *mut _ = &mut MongoHandle::Connection(RwLock::new(
+            Connection::with_state(env_handle, ConnectionState::Allocated),
+        ));
+
+        let mut handle: *mut _ = &mut MongoHandle::Statement(RwLock::new(Statement::with_state(
+            std::ptr::null_mut(),
+            StatementState::Allocated,
+        )));
+        let handle_ptr: *mut _ = &mut handle;
+        assert_eq!(
+            SqlReturn::SUCCESS,
+            SQLAllocHandle(
+                HandleType::Stmt,
+                conn_handle as *mut _,
+                std::mem::transmute::<*mut *mut MongoHandle, *mut Handle>(handle_ptr),
+            )
+        );
+        assert_eq!(
+            StatementState::Allocated,
+            (*handle).as_statement().unwrap().write().unwrap().state
+        );
+        assert_eq!(
+            1,
+            (*conn_handle)
+                .as_connection()
+                .unwrap()
+                .read()
+                .unwrap()
+                .statements
+                .len()
+        );
+        assert_eq!(
+            SqlReturn::SUCCESS,
+            SQLFreeHandle(
+                HandleType::Stmt,
+                std::mem::transmute::<*mut MongoHandle, Handle>(handle),
+            )
+        );
+        assert_eq!(
+            0,
+            (*conn_handle)
+                .as_connection()
+                .unwrap()
+                .read()
+                .unwrap()
+                .statements
+                .len()
+        );
+    }
+}
+
+#[test]
+fn invalid_free() {
```

Comment:
Not a big deal, but is it possible to test INVALID_HANDLE for HandleType::Env for the sake of test coverage? Would it work with passing it a null pointer somehow?

Reply:
null pointer would segfault. I’ll think of something